cmd/flow-dps-live/main.go

```diff
@@ -393,76 +392,59 @@ func run() int {
-	failed := make(chan struct{})
-	ctx, cancel := context.WithCancel(context.Background())
-	go func() {
-		follow.Run(ctx)
-	}()
-	go func() {
-		start := time.Now()
-		log.Info().Time("start", start).Msg("Flow DPS Live Indexer starting")
-		err := fsm.Run()
-		if err != nil {
-			log.Warn().Err(err).Msg("Flow DPS Live Indexer failed")
-			close(failed)
-		} else {
-			close(done)
-		}
-		finish := time.Now()
-		duration := finish.Sub(start)
-		log.Info().Time("finish", finish).Str("duration", duration.Round(time.Second).String()).Msg("Flow DPS Indexer stopped")
-	}()
-	go func() {
-		log.Info().Msg("Flow DPS Live Server starting")
-		api.RegisterAPIServer(gsvr, server)
-		err = gsvr.Serve(listener)
-		if err != nil && !errors.Is(err, http.ErrServerClosed) {
-			log.Warn().Err(err).Msg("Flow DPS Server failed")
-		}
-		log.Info().Msg("Flow DPS Live Server stopped")
-	}()
-	go func() {
-		if !metricsEnabled {
-			return
-		}
 
-		log.Info().Msg("metrics server starting")
-		server := metrics.NewServer(log, flagMetrics)
-		err := server.Start()
-		if err != nil {
-			log.Warn().Err(err).Msg("metrics server failed")
-		}
-		log.Info().Msg("metrics server stopped")
-	}()
-
-	// Here, we are waiting for a signal, or for one of the components to fail
-	// or finish. In both cases, we proceed to shut down everything, while also
-	// entering a goroutine that allows us to force shut down by sending
-	// another signal.
-	select {
-	case <-sig:
-		log.Info().Msg("Flow DPS Indexer stopping")
-	case <-done:
-		log.Info().Msg("Flow DPS Indexer done")
-	case <-failed:
-		log.Warn().Msg("Flow DPS Indexer aborted")
-	}
-	go func() {
-		<-sig
-		log.Warn().Msg("forcing exit")
-		os.Exit(1)
-	}()
-
-	// We first stop serving the DPS API by shutting down the GRPC server. Next,
-	// we shut down the consensus follower, so that there is no indexing to be
-	// done anymore. Lastly, we stop the mapper logic itself.
-	gsvr.GracefulStop()
-	cancel()
-	<-follow.NodeBuilder.Done()
-	err = fsm.Stop()
+	ctx, cancel := context.WithCancel(context.Background())
+	metricsSrv := metrics.NewServer(log, flagMetrics)
+
+	err = engine.New(log, "Flow DPS Live", sig).
+		Component(
+			"api",
+			func() error {
+				api.RegisterAPIServer(gsvr, server)
+				// FIXME: Check whether it makes sense to check for the ErrServerClosed.
+				return gsvr.Serve(listener)
+			},
+			func() {
+				gsvr.GracefulStop()
+			},
+		).
+		Component(
+			"follower",
+			func() error {
+				follow.Run(ctx)
+				return nil
+			},
+			func() {
+				// Cancel the context given to the consensus follower.
+				cancel()
+				// Wait for it to have stopped gracefully.
+				<-follow.NodeBuilder.Done()
+			},
+		).
+		Component(
+			"mapper",
+			func() error {
+				return fsm.Run()
+			},
+			func() {
+				fsm.Stop()
+			},
+		).
+		Component(
+			"metrics",
+			func() error {
+				if !metricsEnabled {
+					return nil
+				}
+				return metricsSrv.Start()
+			},
+			func() {
+				metricsSrv.Stop()
```

Comment:
I'm assuming this is idempotent, but it would be cleaner to actually only shut it down if it was actually started, and make it _not_ idempotent. By "hiding" errors on stopping stuff that is not running, we actually create surface for issues in the logic that we don't spot, because they silently don't fail.